### ISSUE TITLE
Interested in conanfile.py ?

### ISSUE THREAD

**Issue by dotChris90** [eclipse-iceoryx/iceoryx#764]
Hello iceoryx group, 

great work! never saw such impressive performance powered project. 
I am currently working on a conanfile.py (https://github.com/dotChris90/iceoryx/blob/master/conanfile.py) for iceoryx so that cross compilation and dependency management works better "out of the box" (plus was thinking that depending projects like eclipse DDS would get benefit from it). 

Today was first "try out how good it works" but the cmake build is very good and self explaining so was just some lines to add in conanfile and had first shoot. 

My question would be now if you guys interest in a conan package and would accept such pull requests (after finished my work) or if dependency / package management is not in focus of this project now (then i would maybe better contribute to conan-center).

**Comment by elfenpiff** [comment]
@dotChris90 I think this is a great idea!

@dkroenke but then I think we need a conan target in our CI.

**Comment by dotChris90** [comment]
@elfenpiff  thanks for the positive feedback. I think I now created a proper conanfile that automatically takes the VERSION file for for proper versioning, adapting the CMakeFiles on package creation (so we add conan based CMAKE commands to the CMakeFiles without touching them in repo - so other people still can use the CMakeFiles without conan), the git clone of tomlcpp is "outcomment" on creation time of package (we add tomlcpp as dependency) and we have 2 options currentently. option 1 - as static or dynamic libs and option 2 with or without tomlcpp config. I tried cross compilation with 2 Yocto based conan profiles - also cross compilation seems to work nice.

**Comment by dotChris90** [comment]
An extremly positive effect of conan is the easy cross compilation via profiles - i think what would be awesome would be profiles for default linux, raspberry pi (because easy to buy "embedded device" ) and QNX.

**Comment by dotChris90** [comment]
Only cross compilation problem i got was related to the linaro toolchain https://releases.linaro.org/components/toolchain/binaries/7.5-2019.12/arm-linux-gnueabihf/ which is popular for several reasons ... "fatal error: sys/acl.h: No such file or directory" but if i am not totally wrong that is just because the linaro toolchain just comes with "/usr/include/linux/posix_acl.h" and not with "sys/acl.h" - so it's toolchain problem and not project.

**Comment by dotChris90** [comment]
For information and update reason - to get a package into the conan-center (so the main Package store / artifactory ) you have to do a PULL Request on their repo - so the conanfile.py has to be part of their repo. 
https://github.com/conan-io/conan-center-index/issues/5418

This means in case conanfile shall be also part of this repo (for test purpose) we can put it here - but still for getting it into conan center we need to have a 2nd one on their repo.

**Comment by dkroenke** [comment]
@dotChris90 Thanks for the contribution, having a conanfile in our repository gives other projects the chance for easier integration with iceoryx. Especially the deployment with different architectures and compiler can be done fast with that.

Publishing iceoryx into conan-center is also interesting, as i understood you correctly we have then to maintain two conanfiles and also to take care that the package in conan-center is up-to-date for every release. I just took a rough look at the conan-center docs but it seems that there is no automatic workflow updating the conan file and config in conan-center or?

**Comment by dotChris90** [comment]
@dkroenke - yes that's correct. Honestly i prefer the conanfile in project repo approach since the project knows best "what is my target platforms, my target ECUs, my target OS, etc." and so project can better validate the "cross compilation" process plus detect errors earlier (e.g. adding dependencies which contains code a platform does not support - and then see "oh yes here we need to add an abstraction) 

unfortunately - their official "process to add conan packages to the conan center" looks the following : forking from https://github.com/conan-io/conan-center-index then adding ur conanfile to their folder structure in folder recipes and then start PR. They will have a review and after several "hock-ups" are done. 

But to sum it up yes you are right. it would requires at least that the "source" method of these conanfiles look different. Plus it would also means : with every release we would have to adapt the conanfile. in best case (if the cmake parameters dont change or in general the "normal" build ) you just have to add the sources to a conandata.yml file --> like https://github.com/conan-io/conan-center-index/blob/master/recipes/tinyxml2/all/conandata.yml. 

So release iceoryx --> checking if conanfile still working --> add the source to conandata.yml in conan-center github --> do PR --> get into the conan-center.

**Comment by dotChris90** [comment]
I would suggest the following : 

- i already requested cpptoml be part of the conan center index (https://github.com/conan-io/conan-center-index/pull/5428) 
- after this is done and the iceoryx dont mind i would create a conanfile.py on their repo for iceoryx (so outside the iceoryx repo but on their repo )
- after the merge request is done i will take this conanfile and adapt it so it matches the iceoryx repo - then we can think about the cross compilation tests on CI servers. 

So my idea is now : first bring iceoryx in the conan center - i guess especially your partners "cyclonedds" or general spoken "ROS" would get a lot of benefits from it. After this is archieved cross compilers on CI in iceoryx repo would be the target.

**Comment by dkroenke** [comment]
@dotChris90 This sounds like a good plan!
Starting with having iceoryx 1.0 on conan-center is a good beginning. My proposal would be that we maybe maintain only major releases and bugfixes only on demand on conan-center. I'm happy to review then the PR on conan-center for correctness. Due to the fact that we already maintain the iceoryx packages on rosdistro for ROS 2, we would be happy about volunteers who maintain the conan file (with the support of the iceoryx committers of course). Maybe you would jump in? ;-)

An idea can be that you could write a little tutorial on how to upgrade the iceoryx version in conan-center, if i saw it right it is mainly updating the `conandata.yml` in the recipe.

**Comment by dotChris90** [comment]
@dkroenke  i would be happy to make this great project more visible to other projects ;) also would create an issue on cyclonedds - if they are interested. i guess yes. 

But don't worry ;) i would take care of the conanfile.py related stuff. - iceoryx dependency cpptoml - already nearly part of the conan-center (tests green, automatic checks green, 1/3 reviewer already checked). 

Then next step would be iceoryx in version 1.0.0.

**Comment by dkroenke** [comment]
@dotChris90 In the name of the iceoryx community i say thank you! I will assign this issue here to your username so others know that you are working on that.

**Comment by dotChris90** [comment]
First proposal - functional working - taking release 1.0.0 from github - https://github.com/dotChris90/conan-center-index/blob/adding-iceoryx/recipes/iceoryx/all/conanfile.py - it's functional - hooks, pep8, etc. not yet done - have to look more if functional on different settings - cpptoml request still open. test_package not yet implemented --> best we put examples here so test they are functional.

**Comment by dkroenke** [comment]
@dotChris90 
I just took a rough look over the file, i have currently only minor remarks:
- homepage can be "https://iceoryx.io"
- for the topics i would propose to add "Middleware"
- regarding the options i would propose to enable `shared` per default because iceoryx is deployed as shared library in ROS 2 and with CycloneDDS. 

One additional note: we plan to add in future the support for https://github.com/rafagafe/tiny-json for roudi configuration.

**Comment by dotChris90** [comment]
Will be done - the "shared=True" as default is "uncommon" - usual they using really static linking as default (which is in my opinion somehow strange but anyway let's see). let's hope their "reviewer" don't see it as "standard" and blocking.

**Comment by elBoberido** [comment]
Maybe we should also think about static builds for conan. With ROS there is this workflow with sourcing `setup.sh`. As far as I know there is no equivalent on the conan side which would make the usage of iceoryx quite cumbersome

**Comment by dotChris90** [comment]
nice point. First of all i can confirm currently the iceoryx "conan package under preparation" is able to export static and dynamic libs - depending on the options - but both builds working fine. A consuming package / project like cyclonedds or ros2  have do write own conanfile and inside there say "options["iceoryx"].shared=False " then they can link against the static libs. but right - this somehow also depends on how "willing they are to integrate conan to their build" 

@elBoberido  - this sourcing setup.sh - it sounds somehow like the cross compilers and compiler flags etc. are sourced as environment variables and then cmake automatically detect everything (at least the flow i know usual). If this is the case - usual conan uses profiles like https://docs.conan.io/en/latest/integrations/cross_platform/android.html instead of source files. benefit is that you easily can say for your deployment : conan create -pr android . , conan create -pr webassembly . , conan create -pr linux_aarch64 . - etc. 

But very strangly i didnt find any "ROS" profile on internet at first shoot. u find nearly profile "suggestions" for nearly anything. WebAssembly, Android, QNX, Windows, Raspberry, etc.

**Comment by dkroenke** [comment]
@dotChris90 I guess there is a misunderstanding here. @elBoberido meant that when you have built ROS 2 with colcon (a buildtool for orchestrating CMake projects) you need to source a generated `setup.sh` to make the ROS 2 application accessible. Under the hood are runtime dependencies for the shared libraries in ROS 2 resolved.
The conan profiles look for me more like compile-time settings that you need to set before the CMake build.

After some thinking about that, it make also for me more sense to build the conan package of iceoryx per default as static lib. Reason is that for example the RouDi daemon needs to know the location of the `iceoryx_posh_roudi` lib when built as shared library. This can be difficult when iceoryx is installed into a custom folder path.

**Comment by elBoberido** [comment]
Yes, that's what I meant. If the libs are not in a known path, the binaries cannot find them

**Comment by dotChris90** [comment]
Ah guys - sorry for the misunderstanding and thanks for the clearification. 

I think for getting the dependencies and place them to correct location - there exist plenty options too. 
- conan install    # can genererated cmake files if required and this cmake files containing the absolute paths of libs 
- conan install -g deploy # can download and install ur dependencies in front of ur current location 

But i think thats something that we should look at the "consuming" side. 

currently the key consumers (which i read out and know) 
- ROS 2
- cyclone dds 

Is this correct? .

**Comment by dotChris90** [comment]
Ok latest status - https://github.com/dotChris90/conan-center-index/blob/adding-iceoryx/recipes/iceoryx/all/conanfile.py the homepage and topic is adapted. the shared still false default - so static libs. 

The ROS2 thing i could check and see how they best consuming. 

Tinyjson - good you pointed out. I will try to bring them to the conan center index ^^ actually i am suprised - tinyxml is part and tinyjson not.

**Comment by dotChris90** [comment]
Got your point - the ros thing i can say - it's more a "detection / consumption" thing - which need to know more how to use conan with colcon (it's interesting to see how many dependency managers exist in OSS for C++ xD). 

The other thing - integration --> i made a very very basic conanfile for cyclone dds - https://github.com/dotChris90/cyclonedds/blob/iceoryx/conanfile.py i can say - for the default profile no problem looks fine. iceoryx was found - build was sucessful. so i green light. after iceoryx is in Conan Center - we can ask them if they interested - at least they know conan.

**Comment by dotChris90** [comment]
Ok guys - i think the conanfile.py for conan center did a good step forward. https://github.com/dotChris90/conan-center-index/blob/adding-iceoryx/recipes/iceoryx/all/conanfile.py

@dkroenke i think the meta data are now ok too. 
 - name = "iceoryx"
 - license = "Apache-2.0"
 - homepage = "https://iceoryx.io/"
 - url = "https://github.com/conan-io/conan-center-index"
 - description = "Eclipse iceoryx - true zero-copy inter-process-communication"
 - topics = ("Shared Memory", "IPC", "ROS", "Middleware")
 - settings = "os", "compiler", "build_type", "arch"

The options are deduced from the cmake options (DDS Gateway, Tests and examples are not options).
I would like to hear if this options are ok - the default options are also taken from the cmake files. 

The only ... drawback currently this conanfile.py has is mostly related to exported artifacts : 
- it doesnt export the cmake files to "consuming packages / project"
- the "etc" config file is moved to bin folder and the share is deleted (as my local conan center hooks suggested)
- consuming packages / projects must explicit know the libs for linking (see test consuming CMakeFile which is taken from iceoryx examples https://github.com/dotChris90/conan-center-index/blob/adding-iceoryx/recipes/iceoryx/all/test_package/CMakeLists.txt) 
- that's proof of working but still feels not so elegant as with find_package and linking against 1,2 targets (just like in iceoryx repo examples)
- the Conan Center forbits the export of cmake files for the following reason : packages shall be consumed independent of building system 
- usual conanfile.py shall contain a package_info method and here you place several meta data which are the basis for cmake - generators etc. (Conan generates cmake files from meta information but also visual studio or autoconfig build systems - that's how conan integrate into the cycle - generated cmake files are taken into your CMakeFiles.txt)

**Comment by dkroenke** [comment]
Hi,
yes for now we can stay with the option set you already defined in the conanfile (TOML and shared libs option).
From my perspective the following options are not necessary:
- `fPIC` should not be set by the user, this option is set in our CMake files by enabling `POSITION_INDEPENDENT_CODE`
- `c_bind` i think we can deploy the C-Binding automatically together with posh and utils. It's a quite small lib and does not hurt.
- `strict_one2many` This can maybe discussed but from my point of view it is not necessary because we do not want to limit iceoryx here.
- `sanitize` I'm not sure but is it common sense to deploy packages with compiled sanitizer flags in conan? This could lead to incompatibilities when the end-user use another version of the sanitizer than we are deploying.

Other flags like like DDS_GATEWAY or something else i would also exclude in the beginning.

Regarding the CMake file in https://github.com/dotChris90/conan-center-index/blob/adding-iceoryx/recipes/iceoryx/all/test_package/CMakeLists.txt: The `iceoryx_posh_roudi` library is not necessary to link against. It's purpose is to build the roudi middleware daemon.
But it's a good idea to test the roudi build with the conan package, the CMake commands for building it are here: https://github.com/eclipse-iceoryx/iceoryx/blob/e3790ded5031efa23cb3a5a40e0869c1acb07cb2/iceoryx_posh/CMakeLists.txt#L342
with this sourcefile: https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_posh/source/roudi/application/roudi_main.cpp

Maybe i can test the conanfile locally to see which files are deployed and if they are complete.

**Comment by dotChris90** [comment]
@dkroenke @elfenpiff  thanks you two for the considerings and feedback. I minimized the options to shared and toml_config. Because i have to admit your points makes absolute sence. 

- lower options - lower misunderstanding
- sanitize ... yes ... i saw the options and though maybe useful 
- but @elfenpiff  is correct, the package should be easy to use and a sanitize options could lead to "confussion" or "misunderstanding" to consuming packages 
- if people are seriously interested into sanitize they probably have enough experience to get the github repo of iceoryx and do the build on their own 
- fPIC i was suprised that my "local conan center hooks" dont show the missing fPIC options as red (honestly never saw packages that dont have fPIC options so i though it's a standard) 


Related to the CMakeLists in test_package folder 
- the package can now be consumed via the find_package in cmake 
- it reduces the libs a consumer package has to care for linking drastically (just need to target_linked_library against posh)
- i currently looking if there exist sth like an alias because current consumption looks like "iceoryx::posh" and in examples its "iceoryx_posh::icoryx::posh" 
- i think i found all internal dependencies (posh needs utils and utils needs platform such things) and system dependencies (like acl and pthreads)  - they are now in the package_info method

Some hints to local building @dkroenke : 
-  if you want to see what artifacts are deployed - then it's the same like in cmake when installing 
- if you curious about the package itself then after cloning the conan center repo of me and switched to the right branch
```
cd conan-center-index/recipes/iceoryx/all
conan create . 1.0.0@
```
(the reason here : version will be set by their servers so version is not a part of the conanfile but for local try out have to set it via command line)

**Comment by dotChris90** [comment]
ok iceoryx team - current status of https://github.com/dotChris90/conan-center-index/blob/adding-iceoryx/recipes/iceoryx/all/conanfile.py 

- options now just shared and toml-config 
- c bindings are always taken 
- findIceoryx cmake files are generated by conan for consuming project 
- aliases exist in form of "iceoryx_posh::iceoryx_posh" is alias for "iceoryx::posh
- with this help examples from iceoryx can be taken without big change
- only difference : i was not able to find any way to get an find_package alias so find_package must be always "find_package(iceoryx REQUIRED)" 
- package checks if compiler supports C++14 and target is "not windows" 
- even C binding support examples works without problem - !yes! https://github.com/dotChris90/conan-center-index/blob/adding-iceoryx/recipes/iceoryx/all/test_package/icedelivery_in_c/CMakeLists.txt
- examples are functional (started Roudi and started examples)
- hint : i think we should just check if compiling and linking works (so consumption) Conan Center always mention that the Github Conan-Center shall be focused on check if package is working - and not if functions are correct (so no unit tests and no execution of tests)
-  i removed some lines from CMakeFiles of examples for simplification (please simple comment here if you find i removed too much)

I would like to know if you guys are ok with this and you think we could maybe starting the PR - after the PR is started the conan hooks will be started and 3 reviewers are required. let's see. i think the hooks and package creation will be interesting if all runs ok. 
But quite excited - looks at least we close to deployment.

**Comment by elfenpiff** [comment]
@dotChris90 I think this all sounds awesome and I am really looking forward to your PR, so please create it!

At the moment it is hard for me to answer if you cleaned up too good but the PR diff will show us. Either way, thanks for the great work.

**Comment by dotChris90** [comment]
Ok during the day i will create PR and simple let's see what is coming. :D anyway if we notice after that we must add sth - nobody forbit us "extending" existing conan packages. it's always just a PR in Github. so let's see and try ;) will update you when it's in conan center.

**Comment by dotChris90** [comment]
ok first try summery : 

- windows builds were skipped (yes - at least that worked)
- MAC build successful but tests failed (because the C++ Standard detection were not totally correct on my side) 
- Linux build failed because the compilation is inside containers without ACL 
- first step before go on --> i bring ACL into their Center(is quite ok since normal config build and small dependency) 
- iceoryx then would have acl as a dependency if target os is linux 
- our benefit would be --> then we could take e.g. a cross compiler toolchain like linaro and have a profile for cross compilation checks on our side 
- PR present here https://github.com/conan-io/conan-center-index/pull/5541 
- i keep you updated ^^

**Comment by dotChris90** [comment]
@dkroenke  thanks for review - the whitespace i seriously didnt notice -.- i will keep you update guys

**Comment by dotChris90** [comment]
ok - just an update because already some time passed. We had the issues that the containers are much more restricted in their jenkins builds. actually acl was not present so first adding acl to conan center. but for this libattr was required. 

To sum it up - the packages are now present in Conan Center - so next steps will be now Iceoryx inside their containers. 
I will inform you when we reaching the goal.

**Comment by dotChris90** [comment]
Ok - validated in containers and currently checking why for projects linked the old ABI (see https://docs.conan.io/en/1.36/howtos/manage_gcc_abi.html) so the libstdc++11 test_packages linking fails. I thought there exist sth like an adaption for "backwards compatibility". But Gentlemen - it looks like iceoryx can be compiled and linked in their restricted containers - beginning from gcc6! for linux. 

My small question here would be - official in ur instructions u pointing out that u need gcc7 - should we forbid gcc6 use (because it leads to many many warnings - no errors) or shall we simple "print a warning - "be careful - iceoryx was designt for gcc7 and higher"? 

I will this week adapt the other changes and finally trigger again the jenkins workers.

**Comment by dkroenke** [comment]
Regarding the compiler support we also have limited support for GCC 5.4 and a CI Job running for it: https://github.com/eclipse-iceoryx/iceoryx/blob/4b2736be424aceab7888ace5741ba24ecd5b3614/.github/workflows/experimental_build.yml#L52
We did that to ensure compatibility with QNX 7.0 but we will deprecate this support in the mid-term future.

We give for compiler version lower than gcc 7 only the support that it compiles and run tests without errors.Warnings are tolerated there. In versions from 7 on we do not tolerate compiler warnings anymore and our quality levels apply: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#quality-levels

I would not forbid older compiler to not exclude contributors who want to use them, they can every time create a ticket here and we can discuss a solution if compilation fails. We also test against the latest clang compiler version.

**Comment by dotChris90** [comment]
thanks here! currently i will print warning when gcc < 7. 

I want to share the information with you that the package was able to built on their jenkins servers for MAC, Linux and Windows. The restrictions is mostly related to the so called libstdc++11 thing. Since gcc 5 exist 2 different ABIs in linux - which was necessary because C++11 standard forced the library implementers (so the people who implement the lib) to make behaviour changes (especially in string class etc.) so there exist different functions (either you choose ABI new or old) and iceoryx seems to work proper with libstdc++11 (new ABI) - but that's not sth iceoryx has to care.

**Comment by dotChris90** [comment]
Two question i would like to ask are the following : 

- i read Windows support is just with Visual Studio compiler > 16 (Visual Studio 2019 and higher) right? 
- is Iceoryx also buildable with "shared libs" (so DLLS instead of *.libs ) in windows?

**Comment by dkroenke** [comment]
@dotChris90 
We are currently building the windows target on the following CI environment:
https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md#visual-studio-enterprise-2019
This is the windows version where we give support.
Regarding the shared lib build i need to check it out when I have access next time to windows PC.

**Comment by dotChris90** [comment]
``@dkroenke i am sorry it took so long - but the PR is closed. Iceoryx 1.0.0 is part of CONAN Center and so reachable from conan dependency manager - hope later packages will be faster. https://conan.io/center/iceoryx 

Actually i have some technical depth to pay to them (3 reviewer gave green light - one reviewer pleased me to do some changes - so PR merged - i will open an other one). 

The only one that is little bit weird to me is : the compiler must be forced to link against libstdc++11 (the new ABI) but i cannot find a reason. that's maybe one thing i also would like to check later. 

Currently developer can get iceoryx artifacts directly via : 
``
conan install -s compiler.libcxx=libstdc++11 -g deploy iceoryx/1.0.0@_/_ --build=missing > /dev/null
ls iceoryx/lib/
``
This command will compile or cross compile the package and "spit it out" in front of you. it is a good option if you e.g. have embedded linux with it's rootfs and you want to copy the artifacts into the rootfs (simple iterate over the lib folders and copy).

for sure : if you want to integrate iceoryx into a project  ur project needs a conanfile.py or conanfile.txt (*.txt normally is used if you want to collect dependencies but not package your project).

**Comment by dotChris90** [comment]
just small question - just to be sure. :D i know iceoryx has a company behind - so it's better to ask double. you guys allow PR to iceoryx? 2 things i currently have in my mind. 

1) adding documentation support (so conan besides colcom)
2) dependency  resolution. 

I want to mention that dependency management can be done for a project with a conanfile.txt file (this is pure dependency resolution reason - no packaging - see https://docs.conan.io/en/latest/reference/conanfile_txt.html) - basically conan will compile the dependencies and generate cmakefiles so that your cmakefile can use "find_package" and find it's dependencies. with this thing iceoryx would avoid "git submodule" stuff and so on. 

Since this is a direct impact on the build workflow of iceoryx i just want to mention that there exist this possibility - the packaging still would be done in the conan-center index github via PR.

**Comment by dkroenke** [comment]
@dotChris90 Thank you very much, we appreciate your effort!
I will test the conan workflow in the next weeks and give feedback on this, but will do this in a separate ticket because this one is quite good commented already.
FYI: We recently released a v1.0.1 bugfix release for iceoryx, so if you create a clean-up PR on conan center would you then may add the new version. You can add me as reviewer then.

Since there are more open points like the ARM64 build, i would propose to do this in a separate ticket.
For now we close the ticket here because the original scope to bring iceoryx into conan-center is fullfilled.

**Comment by elBoberido** [comment]
@dotChris90 thanks very much for your effort 👍
